src/meta-srv/src/service/heartbeat.rs

```diff
@@ -48,10 +51,10 @@ impl heartbeat_server::Heartbeat for MetaSrv {
                 let mut quit = false;
                 match msg {
                     Ok(req) => {
-                        let role = req.header.as_ref().map_or(0, |h| h.role);
+                        let role = req.header.as_ref().map_or(Role::Datanode, |h| h.role());
                         if pusher_key.is_none() {
-                            if let Some(peer) = &req.peer {
-                                let key = format!("{}-{}", role, peer.id,);
+                            if let Some(node_id) = get_node_id(&req) {
+                                let key = format!("{}-{}", role as i32, node_id,);
```

Comment:
> - assign unique id for frontend, but not for datanode
> - using etcd to generate unique ids

Heartbeat information can only be processed by the meta leader, and most of the mailboxes of meta and fronend are broadcast. So I still think generating frontend_node_id using atomic is ok. 😃

Reply:
> - reject requests that doesn't have role in its header
I think it is ok. PTAL @fengjiachun

Reply:
> > * reject requests that doesn't have role in its header
> 
> I think it is ok. PTAL @fengjiachun

+1

Reply:
I think the default value should not be Datanode, as this will affect the resource scheduling of meta. Therefore, I agree with @MichaelScofield's proposal that we should reject streams without empty role .